rclpy/rclpy/executors.py

```diff
@@ -745,24 +745,45 @@ class SingleThreadedExecutor(Executor):
     def __init__(self, *, context: Optional[Context] = None) -> None:
         super().__init__(context=context)
 
-    def spin_once(self, timeout_sec: Optional[float] = None) -> None:
+        self._futures = []
+
+    def _spin_once_impl(
+        self,
+        timeout_sec: Optional[float] = None,
+        wait_condition: Callable[[], bool] = lambda: False
+    ) -> None:
         try:
-            handler, entity, node = self.wait_for_ready_callbacks(timeout_sec=timeout_sec)
+            handler, entity, node = self.wait_for_ready_callbacks(
+                timeout_sec, None, wait_condition)
         except ShutdownException:
             pass
         except TimeoutException:
             pass
+        except ConditionReachedException:
+            pass
         else:
             handler()
             if handler.exception() is not None:
                 raise handler.exception()
 
+            self._futures.append(handler)
+            # make a copy of the list that we iterate over while modifying it
+            # (https://stackoverflow.com/q/1207406/3753684)
+            for future in self._futures[:]:
+                if future.done():
+                    self._futures.remove(future)
+                    future.result()  # re-raise any exceptions
```

Comment:
Do we need to re-raise the exception here again? the exception comes from user callback is already managed above for `SingleThreadedExecutor` case?

Reply:
I *think* we need this.  The reason is that the handler exception above is only for the current handler; if we had other futures in this list, then it may be possible that one of them is "done" because of an exception, and we need to notify the user about that.

Reply:
I guess it's less so a re-raise as it is just raising the exception? (Since if the exception was raised in line 767, the handler wouldn't get added to the future at all?)

Reply:
I may be mistaken, but do we need to have `SingleThreadedExecutor::_futures`? handler is executed on this thread in line 765 with checking the exception to re-raise it for user application every?

Reply:
> I guess it's less so a re-raise as it is just raising the exception? (Since if the exception was raised in line 767, the handler wouldn't get added to the futures list at all?)

Yeah, I think that is correct.  I copied that comment from the `MultiThreadedExecutor` implementation, but it is indeed just a "raise".  I'll update the comment to reflect that.

Reply:
> I may be mistaken, but do we need to have `SingleThreadedExecutor::_futures`? handler is executed on this thread in line 765 with checking the exception to re-raise it for user application every?

Actually, this is a good question.  In the `SingleThreadedExecutor` case, we are executing the handler directly (not submitting it to a ThreadPool to "eventually" be run).  So there is really no reason to have `self._futures` at all; we can just run the handler, then raise any exceptions immediately by just running `handler.result`.  I'll try that out.

Reply:
OK, yeah, that seems to work.  So that is now done in fd30442.